src/helm/benchmark/scenarios/melt_scenarios.py

```diff
@@ -343,3 +351,947 @@ def __init__(self, **kwargs):
+        rule_attributes, attributes_shuffled = (
+            attributes_shuffled[:n_rule_attributes],
+            attributes_shuffled[n_rule_attributes:],
+        )
+        rules.append(
+            MELTLanguageRule(
+                subject=rule_subject,
+                subject_category=subject_category,
+                specifier_type="một",
+                condition=rule_attributes[:-1],
+                condition_conjunction=random.choice(["và", "hoặc"]),
+                consequent=rule_attributes[-1],
+            )
+        )
+    return rules
+
+
+def generate_test(
+    attribute_groups: Dict[str, List[str]],
+    subject: str,
+    subject_category: str,
+    rules: List[MELTLanguageRule],
+    use_specific_attributes: bool,
+    p_consequenceless=0.1,
+) -> Tuple[MELTLanguageFact, List[MELTLanguageRule], MELTLanguageFact]:
+    """Generates a test case given a set of rules, i.e. a statement about the subject from which something
+    can be potentially deduced given the rules. We include an argument, p_consequenceless, to re-roll with
+    some probability if the generated fact does not allow anything to be determined.
+    """
+
+    # The generic attributes which the test fact will assign to the subject
+    test_attributes: List[str] = random.sample(list(attribute_groups.keys()), 2)
+    # The specific versions of the test attributes
+    test_attributes_specific: List[str] = [
+        random.choice(attribute_groups[subcondition]) for subcondition in test_attributes
+    ]
+    test_consequents: List[str] = []  # The attributes implied by the test attributes and rules
+    test_rules_used: List[MELTLanguageRule] = []
+    for rule in rules:
+        if rule.consequent in test_attributes:
+            continue
+        if rule.condition_conjunction == "và":
+            if set(rule.condition).issubset(test_attributes):
+                test_rules_used.append(rule)
+                test_consequents.append(rule.consequent)
+        elif rule.condition_conjunction == "hoặc":
+            if not set(rule.condition).isdisjoint(test_attributes):
+                test_rules_used.append(rule)
+                test_consequents.append(rule.consequent)
+    if len(test_consequents) == 0 and random.random() > p_consequenceless:
+        return generate_test(
+            attribute_groups, subject, subject_category, rules, use_specific_attributes, p_consequenceless
+        )
+
+    test_fact: MELTLanguageFact = MELTLanguageFact(
+        subject,
+        subject_category,
+        specifier_type="cái",
+        specific_attributes=test_attributes_specific,
+        generic_attributes=test_attributes,
+        use_specific_attributes=use_specific_attributes,
+    )
+
+    target_fact: MELTLanguageFact = dataclasses.replace(
+        test_fact,
+        specific_attributes=test_consequents,
+        generic_attributes=test_consequents,
+    )
+
+    return test_fact, test_rules_used, target_fact
+
+
+class MELTSRNScenario(Scenario):
```

Comment:
This scenario has a lot of code. Please move it and its helper functions to its own file `melt_srn_scenario.py`.